rules/nette/src/Rector/Assign/ArrayAccessAddRouteToAddRouteMethodCallRector.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Nette\Rector\Assign;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\ArrayDimFetch;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\Variable;
+use Rector\Core\Rector\AbstractRector;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Nette\Tests\Rector\Assign\ArrayAccessAddRouteToAddRouteMethodCallRector\ArrayAccessAddRouteToAddRouteMethodCallRectorTest
+ */
+final class ArrayAccessAddRouteToAddRouteMethodCallRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Change magic array access add Route, to explicit $routeList->addRoute(...) method',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+use Nette\Application\Routers\Route;
+use Nette\Application\Routers\RouteList;
+
+class RouterFactory
+{
+    public static function createRouter()
+    {
+        $routeList = new RouteList();
+        $routeList[] = new Route('<module>/<presenter>/<action>[/<id>]', 'Homepage:default');
+        return $routeList;
+    }
+}
+CODE_SAMPLE
+,
+                    <<<'CODE_SAMPLE'
+use Nette\Application\Routers\RouteList;
+
+class RouterFactory
+{
+    public static function createRouter()
+    {
+        $routeList = new RouteList();
+        $routeList->addRoute('<module>/<presenter>/<action>[/<id>]', 'Homepage:default');
+        return $routeList;
+    }
+}
+CODE_SAMPLE
+                ),
+
+            ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [Assign::class];
+    }
+
+    /**
+     * @param Assign $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $this->isAssignOfRouteToRouteListDimFetch($node)) {
+            return null;
+        }
+
+        /** @var ArrayDimFetch $arrayDimFetch */
+        $arrayDimFetch = $node->var;
+
+        /** @var Node\Expr\New_ $newRoute */
+        $newRoute = $node->expr;
+
+        return new MethodCall($arrayDimFetch->var, 'addRoute', $newRoute->args);
+    }
+
+    private function isAssignOfRouteToRouteListDimFetch(Assign $assign): bool
+    {
+        if (! $assign->var instanceof ArrayDimFetch) {
+            return false;
+        }
+
+        if (! $this->isObjectType($assign->expr, 'Nette\Application\Routers\Route')) {
+            return false;
+        }
+
+        $arrayDimFetch = $assign->var;
+        if (! $arrayDimFetch->var instanceof Variable) {
+            return false;
+        }
+
+        return $this->isObjectType($arrayDimFetch->var, 'Nette\Application\Routers\RouteList');
```

Comment:
The class and method should be passed via `configure()` method, so this rule can be used as generic one.

```php
$services->set(ArrayAccessAddRouteToAddRouteMethodCallRector::class)
    ->call('configure', [[
        ArrayAccessAddRouteToAddRouteMethodCallRector::TYPE_TO_METHOD_CALL => [
           'Nette\Application\Routers\Route' => 'addRoute'
        ]
    ]]);
```

Is the second class 'Nette\Application\Routers\RouteList' really neccessary or can we remove it?

Reply:
I'm not sure now. I have to go deeper :) there are at least two classes with method addRoute() but I don't know if both can be used in createRouter method...

But let assume that you have router which just adds Routes to some array and then iterates over this array and adds it to RouteList

Something like this:
```
public function createRouter()
{
    $routeArray = [];
    $routeList = new RouteList();

    $routeArray[] = new Route('<module>/<presenter>/<action>[/<id>]', 'Homepage:default');

    foreach ($routeArray as $route) {
        $routeList[] = $route;
    }
    return $routeList;
}
```
With ignoring RouteList, our code will call ->addRoute on $routeArray

Reply:
I see. Thanks for the code.
The right part must be checked for `New_` node instance.

Reply:
check added

Reply:
This needs to be resolved. Rule must be configurable

Reply:
This needs to be resolved. Rule must be configurable

Reply:
Only time will tell. This abstraction is obvious (class and method name), so better do it now when we have focus, then in 6 month create duplicated rule. Yes that happens :)

Reply:
OK I will try something. If I have 3+ config strings (listClass, itemClass, listMethod) I will need some ValueObject right?

Reply:
Thank you :+1: 

> I will need some ValueObject right?

Yes, that would be way to go for 3+ values to check

Reply:
Thank you :+1: 

> I will need some ValueObject right?

Yes, that would be way to go for 3+ values to check

Reply:
Generic rule is created, but probably some renamings should be done...